File: flutter_client/lib/services/api_service.dart
import 'dart:convert';

import 'package:flutter_client/app_config.dart';
import 'package:flutter_client/models/auth_models.dart';
import 'package:flutter_client/models/signal_models.dart';
import 'package:http/http.dart' as http;

class ApiException implements Exception {
  final String message;
  final int? statusCode;

  ApiException(this.message, {this.statusCode});

  @override
  String toString() => message;
}

class UnauthorizedException extends ApiException {
  UnauthorizedException()
    : super('Session login sudah tidak valid.', statusCode: 401);
}

class ApiService {
  Uri _uri(String path) => Uri.parse('${AppConfig.baseApiUrl}$path');

  Future<LoginResponse> login(LoginRequest request) async {
    final response = await http.post(
      _uri('/auth/login'),
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: jsonEncode(request.toJson()),
    );

    final body = _decodeBody(response.body);
    if (response.statusCode >= 400) {
      throw ApiException(
        (body['message'] ?? 'Login gagal').toString(),
        statusCode: response.statusCode,
      );
    }

    return LoginResponse.fromJson(body);
  }

  Future<List<SignalItem>> getSignals(String token) async {
    final response = await http.get(
      _uri('/client/signals'),
      headers: {'Accept': 'application/json', 'Authorization': 'Bearer $token'},
    );

    if (response.statusCode == 401) {
      throw UnauthorizedException();
    }

    final body = _decodeBody(response.body);
    if (response.statusCode >= 400) {
      throw ApiException(
        (body['message'] ?? 'Gagal mengambil sinyal').toString(),
        statusCode: response.statusCode,
      );
    }

    return SignalResponse.fromJson(body).signals;
  }

  Future<void> logout(String token) async {
    await http.post(
      _uri('/auth/logout'),
      headers: {'Accept': 'application/json', 'Authorization': 'Bearer $token'},
    );
  }

  Map<String, dynamic> _decodeBody(String body) {
    if (body.trim().isEmpty) {
      return {};
    }

    final decoded = jsonDecode(body);
    if (decoded is Map<String, dynamic>) {
      return decoded;
    }

    return {};
  }
}

File: android-client/app/src/main/java/com/alima/sinyalsahamindo/data/network/ApiService.kt
package com.alima.sinyalsahamindo.data.network

import com.alima.sinyalsahamindo.data.model.LoginRequest
import com.alima.sinyalsahamindo.data.model.LoginResponse
import com.alima.sinyalsahamindo.data.model.SignalResponse
import retrofit2.http.Field
import retrofit2.http.FormUrlEncoded
import retrofit2.Response
import retrofit2.http.Body
import retrofit2.http.GET
import retrofit2.http.Header
import retrofit2.http.POST

interface ApiService {
    @POST("auth/login")
    suspend fun login(@Body request: LoginRequest): Response<LoginResponse>

    @GET("client/signals")
    suspend fun getSignals(@Header("Authorization") bearerToken: String): Response<SignalResponse>

    @POST("auth/logout")
    suspend fun logout(@Header("Authorization") bearerToken: String): Response<Map<String, Any>>

    @FormUrlEncoded
    @POST("auth/fcm-token")
    suspend fun updateFcmToken(
        @Header("Authorization") bearerToken: String,
        @Field("fcm_token") fcmToken: String
    ): Response<Map<String, Any>>

    @FormUrlEncoded
    @POST("auth/change-password")
    suspend fun changePassword(
        @Header("Authorization") bearerToken: String,
        @Field("old_password") oldPass: String,
        @Field("new_password") newPass: String,
        @Field("new_password_confirmation") confirmPass: String
    ): Response<Map<String, Any>>
}
